File: marketanalysisapp/lib/models/user_alert_model.dart
class UserAlert {
  final int id;
  final int userId;
  final int assetId;
  final String assetSymbol;
  final String assetName;
  final String alertType;
  final double targetPrice;
  final bool isRepeating;
  final String? note;
  final bool isActive;
  final DateTime createdAt;
  final DateTime? lastTriggeredAt;
  final int triggerCount;

  UserAlert({
    required this.id,
    required this.userId,
    required this.assetId,
    required this.assetSymbol,
    required this.assetName,
    required this.alertType,
    required this.targetPrice,
    required this.isRepeating,
    this.note,
    required this.isActive,
    required this.createdAt,
    this.lastTriggeredAt,
    required this.triggerCount,
  });

  factory UserAlert.fromJson(Map<String, dynamic> json) {
    return UserAlert(
      id: json['id'] as int,
      userId: json['userId'] as int,
      assetId: json['assetId'] as int,
      assetSymbol: json['assetSymbol'] as String,
      assetName: json['assetName'] as String,
      alertType: json['alertType'] as String,
      targetPrice: (json['targetPrice'] as num).toDouble(),
      isRepeating: json['isRepeating'] as bool,
      note: json['note'] as String?,
      isActive: json['isActive'] as bool,
      createdAt: DateTime.parse(json['createdAt'] as String),
      lastTriggeredAt: json['lastTriggeredAt'] != null
          ? DateTime.parse(json['lastTriggeredAt'] as String)
          : null,
      triggerCount: json['triggerCount'] as int,
    );
  }
}

class CreateUserAlert {
  final int assetId;
  final String alertType;
  final double targetPrice;
  final bool isRepeating;
  final String? note;

  CreateUserAlert({
    required this.assetId,
    required this.alertType,
    required this.targetPrice,
    this.isRepeating = false,
    this.note,
  });

  Map<String, dynamic> toJson() {
    return {
      'assetId': assetId,
      'alertType': alertType,
      'targetPrice': targetPrice,
      'isRepeating': isRepeating,
      'note': note,
    };
  }
}

class UpdateUserAlert {
  final double? targetPrice;
  final bool? isRepeating;
  final bool? isActive;
  final String? note;

  UpdateUserAlert({
    this.targetPrice,
    this.isRepeating,
    this.isActive,
    this.note,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    if (targetPrice != null) data['targetPrice'] = targetPrice;
    if (isRepeating != null) data['isRepeating'] = isRepeating;
    if (isActive != null) data['isActive'] = isActive;
    if (note != null) data['note'] = note;
    return data;
  }
}

File: MarketAnalysisFrontend/src/app/core/models/user-alert.model.ts
/**
 * User Alert Models
 *
 * Defines DTOs for user-created price alert functionality
 */

// ==================== Request DTOs ====================

/**
 * Data Transfer Object for creating a new price alert
 *
 * Users can create alerts that trigger when price crosses a threshold
 */
export interface CreateUserAlertDto {
  /** ID of the asset to monitor */
  assetId: number;
  /** Alert trigger condition */
  alertType: 'ABOVE' | 'BELOW' | 'REACHES';
  /** Target price that triggers the alert */
  targetPrice: number;
  /** If true, alert can trigger multiple times */
  isRepeating: boolean;
  /** Optional user note about this alert */
  note?: string;
}

/**
 * Data Transfer Object for updating an existing alert
 * All fields are optional - only provided fields will be updated
 */
export interface UpdateUserAlertDto {
  /** New target price */
  targetPrice?: number;
  /** Change repeating behavior */
  isRepeating?: boolean;
  /** Enable or disable the alert */
  isActive?: boolean;
  /** Update the note */
  note?: string;
}

// ==================== Response DTOs ====================

/**
 * Full alert information returned from API
 * Includes creation metadata and trigger history
 */
export interface UserAlertResponseDto {
  /** Unique alert identifier */
  id: number;
  /** Owner user ID */
  userId: number;
  /** Asset being monitored */
  assetId: number;
  /** Asset symbol (e.g., 'BTC') */
  assetSymbol: string;
  /** Asset full name */
  assetName: string;
  /** Alert trigger condition */
  alertType: string;
  /** Target price threshold */
  targetPrice: number;
  /** Whether alert can trigger multiple times */
  isRepeating: boolean;
  /** Whether alert is currently active */
  isActive: boolean;
  /** User's note about this alert */
  note: string;
  /** Alert creation timestamp */
  createdAt: Date;
  /** Last time this alert triggered */
  lastTriggeredAt?: Date;
  /** Total number of times alert has triggered */
  triggerCount: number;
}
